### ISSUE TITLE
"autossl" proxy spec.  Many connections in CLOSE_WAIT state

### ISSUE THREAD

**Issue by vetkom** [droe/sslsplit#157]
Hello!
SSLsplit SSLsplit 0.5.0-37-gab1c676, build from origin/master, compiled against libevent 2.0.22-stable. Debian GNU/Linux 7.8 (wheezy)
Got problem with autossl. 

I run sslsplit from command line:
` sslsplit -D  autossl   0.0.0.0 8888  -k ../etc/iwdmproxy/ca.key -c ../etc/iwdmproxy/ca.crt`

1) Do one https-request  https://ya.ru using firefox, browser connects through proxy 10.60.5.5:3128
I've got web-page correctly. It's ok.
lsof shows:
```
#lsof -i
firefox   115097    vital   53u  IPv4 703715      0t0  TCP my_host:51388->10.60.5.5:3128 (ESTABLISHED)
sslsplit  115453     root    8u  IPv4 704770      0t0  TCP *:8888 (LISTEN)
sslsplit  115453     root   26u  IPv4 704886      0t0  TCP localhost:8888->my_host:51388 (ESTABLISHED)
sslsplit  115453     root   27u  IPv4 704887      0t0  TCP my_host:51390->10.60.5.5:3128 (ESTABLISHED)
```
But after firefox closed connection lsof -i shows:
```
sslsplit  115453     root    8u  IPv4 704770      0t0  TCP *:8888 (LISTEN)
sslsplit  115453     root   27u  IPv4 704887      0t0  TCP my_host:51390->10.60.5.5:3128 (CLOSE_WAIT)
```
And every https-connection leads to one more file descriptor in CLOSE_WAIT state 

I've try to debug a little bit. I've noticed when calling bufferevent_free_and_close_fd for connections upgraded with bufferevent_openssl_filter_new,  bufferevent_getfd() always returned -1.
```
bufferevent_free_and_close_fd(struct bufferevent *bev, pxy_conn_ctx_t *ctx) {
	evutil_socket_t fd = bufferevent_getfd(bev);
```
	
And this fd = -1 was passed to pxy_ssl_shutdown()  

I've try to build sslsplit also with libevent 2.0.5 but with the same result.
And building sslsplit from from origin/develop didn't fix the issue.

I would appreciate any help.
Thank you in advance!

**Comment by ahalam** [comment]
@droe Migrating to `libevent 2.1.8-stable` fixed this issue. But I had to make a change in `bufferevent_free_and_close_fd`. Get the `ssl` pointer unconditionally (call to `bufferevent_openssl_get_ssl`). I suspect that `ctx->spec->ssl` should be set when the connection is upgraded to SSL but I am not sure. Anyhow, the problem seems to be fixed with these 2 actions.

**Comment by droe** [comment]
@ahalam Thanks - can you confirm or deny whether commit 2b80f257b702eee10a744ce7b1bc989b56e4c526 is sufficient to fix the problem for you, with libevent 2.1.8-stable? I am not sure which version of `bufferevent_free_and_close_fd` you are refering to.

**Comment by droe** [comment]
Assuming fixed, please file a new issue if that is incorrect.
